### ISSUE TITLE
ArrayBoundCheckerV2 false positive if indexer has size_t type

### ISSUE THREAD

**Issue by steakhal** [llvm/llvm-project#44493]
|  |  |
| --- | --- |
| Bugzilla Link | [45148](https://llvm.org/bz45148) |
| Version | trunk |
| OS | Linux |
| Attachments | [Rewritten exploded graphs of all three functions of the example.](https://user-images.githubusercontent.com/6280485/143760464-63a6a09d-68b3-4a1b-ac3e-860fbe26a3d5.gz), [state machine representation of ArrayBoundCheckerV2::checkLocation](https://user-images.githubusercontent.com/6280485/143760465-bbbe5a6d-223b-4668-8df6-e5a4a7c3b45c.gz) |
| CC | @steakhal,@devincoughlin,@haoNoQ |

## Extended Description 

This bug was reported by Loïc Joly.
You can read the original discussion at the cfe-dev archives: http://lists.llvm.org/pipermail/cfe-dev/2020-March/064783.html

The gist of this bug is demonstrated by this example:
```
// clang -cc1 -analyze -analyzer-checker=core,alpha.security.ArrayBoundV2 arrayv2.c
typedef unsigned long long size_t;
const char a[] = "aabbcc";

char f1(size_t len) {
  return a[len+1];
//       ^~~~~~~~
// arrayv2.c:7:10: warning: Out of bound memory access (access exceeds upper limit of memory block)
}
char f2(size_t len) {
  return a[len]; // no-warning
}
char f3(int len) {
  return a[len+1]; // no-warning
}
```

**Comment by steakhal** [comment]
Fixed by de2547329b41
